code/game/objects/items/weapons/twohanded.dm

```diff
@@ -608,3 +609,211 @@
+		update_icon()
+		update_item_state()
+
+///proc for the fuel cost and check and chainsaw noises
+/obj/item/weapon/twohanded/chainsaw/proc/rip_apart(mob/user)
+	if(!active)
+		return
+	reagents.remove_reagent(/datum/reagent/fuel, fuel_used) //remove fuel by fuel_used amount
+	user.changeNext_move(attack_speed) //this is here because attacking object for some reason dont respect weapon attack speed
+	if(reagents.get_reagent_amount(/datum/reagent/fuel) < fuel_used && active) //turn off the chainsaw after one last attack when fuel ran out
+		playsound(loc, 'sound/items/weldingtool_off.ogg', 50)
+		to_chat(user, span_warning("\The [src] shuts off, using last bits of fuel!"))
+		active = FALSE
+		toggle_motor(user)
+		return
+	if(prob(1)) // small chance for an easter egg of simpson chainsaw noises
+		playsound(loc, 'sound/weapons/chainsaw_simpson.ogg', 60)
+	else
+		playsound(loc, 'sound/weapons/chainsawhit.ogg', 100, 1)
+
+///Chainsaw give bump attack when picked up
+/obj/item/weapon/twohanded/chainsaw/equipped(mob/user, slot)
+	. = ..()
+	toggle_item_bump_attack(user, TRUE)
+
+///Chainsaw turned off when dropped, and also lose bump attack
+/obj/item/weapon/twohanded/chainsaw/dropped(mob/user)
+	. = ..()
+	toggle_item_bump_attack(user, FALSE)
+	if(active)
+		active = FALSE
+		toggle_motor(user)
+
+///Chainsaw turn on when wielded
+/obj/item/weapon/twohanded/chainsaw/wield(mob/user)
+	. = ..()
+	if(!.)
+		return
+	playsound(loc, 'sound/weapons/chainsawstart.ogg', 100, 1)
+	toggle_active(FALSE)
+	if(!do_after(user, SKILL_TASK_TRIVIAL, NONE, src, BUSY_ICON_DANGER, null,PROGRESS_BRASS))
+		return
+	toggle_active(TRUE)
+	toggle_motor(user)
+
+///Chainsaw turn off when unwielded
+/obj/item/weapon/twohanded/chainsaw/unwield(mob/user)
+	. = ..()
+	if(!.)
+		return
+	toggle_motor(user)
+
+/obj/item/weapon/twohanded/chainsaw/examine(mob/user)
+	. = ..()
+	. += "It contains [reagents.get_reagent_amount(/datum/reagent/fuel)]/[max_fuel] units of fuel!"
+
+///Refueling with fueltank
+/obj/item/weapon/twohanded/chainsaw/afterattack(obj/target, mob/user, flag)
+	if(istype(target, /obj/structure/reagent_dispensers/fueltank) && get_dist(user,target) <= 1)
+		var/obj/structure/reagent_dispensers/fueltank/saw = target
+		if(saw.reagents.total_volume == 0)
+			to_chat(user, span_warning("Out of fuel!"))
+			return ..()
+
+		var/fuel_transfer_amount = min(saw.reagents.total_volume, (max_fuel - reagents.get_reagent_amount(/datum/reagent/fuel)))
+		saw.reagents.remove_reagent(/datum/reagent/fuel, fuel_transfer_amount)
+		reagents.add_reagent(/datum/reagent/fuel, fuel_transfer_amount)
+		playsound(loc, 'sound/effects/refill.ogg', 25, 1, 3)
+		to_chat(user, span_notice("You refill [src] with fuel."))
+		update_icon()
+
+	return ..()
+
+///extra effect when attacking mob, such as xeno and human
+/obj/item/weapon/twohanded/chainsaw/attack(mob/living/carbon/M as mob, mob/living/carbon/user as mob)
+	rip_apart(user)
+	if(!active) //no extra effect when weapon is off
```

Comment:
as mob? 
also you're leaving a lot of unnecessary comments, a person can read that active doesnt do anything and etc

Reply:
To be completely honest, i took that from rocket sledge, and now i think about it it's kinda pointless, ill remove it.
Tbh i started to add comments to help my brain in testing things, ill clean it up on the irrelevant stuffs.